### PR TITLE
Processing of dataset URLs with the new cache organization

### PR DESCRIPTION
**Pull request by candleindark**
This PR is a new implementation of the processing of dataset URLs with [the new cache organization](https://github.com/datalad/datalad-registry/issues/125) and in support of [the V2 API](https://github.com/datalad/datalad-registry/pull/149). It lays the foundation for handling the race condition in accessing the cache, updating the dataset URLs information periodically, and [auditing Celery actions](https://github.com/datalad/datalad-registry/issues/133). Many of the extractions of information to populate a dataset URL object have been rewritten (re-organized in some cases) to handle edge conditions properly and/or to obtain results in more appropriate datatypes.

### CODE REVIEW

**Comment by candleindark** [comment]
The remaining issues before this PR can be merged are

1. The tests for one reason or another are not using database connection economically.  A possible solution can be found https://stackoverflow.com/questions/66876181/how-do-i-close-a-flask-sqlalchemy-connection-that-i-used-in-a-thread
2. Some small problem with the linter or the setup of it.

**Comment by candleindark** [comment]
> The remaining issues before this PR can be merged are
> 
> 1. The tests for one reason or another are not using database connection economically.  A possible solution can be found https://stackoverflow.com/questions/66876181/how-do-i-close-a-flask-sqlalchemy-connection-that-i-used-in-a-thread
> 2. Some small problem with the linter or the setup of it.

Thanks to @yarikoptic's hint during the weekly meeting, I took another look at how the `flask_app` fixture was setup. The Postgres connection issue is now resolved.

File: datalad_registry/models.py
Comment (yarikoptic) [inline datalad_registry/models.py:29]: So what this `cache_path` be now?  Please describe briefly in a comment above, so it also becomes clear why `34` is a sufficient length.

Comment (candleindark) [reply datalad_registry/models.py:29]: `cache_path` will be a relative path of 

```python
        uuid = uuid4().hex
        path = Path(uuid[:3], uuid[3:6], uuid[6:])
```

The path relative to the directory specified by `DATALAD_REGISTRY_DATASET_CACHE`, the base cache path.

File: datalad_registry/utils/misc.py
Comment (candleindark) [inline datalad_registry/utils/misc.py:18]: @yarikoptic This function generates values to be assigned to `cache_path` of the `URL` model.

File: datalad_registry/models.py
Comment (yarikoptic) [reply datalad_registry/models.py:33]: Great, thank you for the explanation!

We need an an endpoint which , given a `cache_path`  would return a dataset record. And then also a command line entry point for the same, and ideally it should be smart to operate on full path, thus taking the last three path components -- this way user (us) could quickly figure out what dataset a given path corresponds to.

Comment (candleindark) [reply datalad_registry/models.py:33]: It makes complete sense to me to have such a command line entry for us, the operators, to have access to such information conveniently. However, since `cache_path` is of the internal organization of a running instance of datalad-registry. Don't you think it is a bad idea to expose it through the public `API`?

Comment (yarikoptic) [reply datalad_registry/models.py:33]: I don't see a problem with that, in particular if it is just a UUID (random) and you need to know it to get some non 404 response.  What concerns do you have from exposing it?

Comment (candleindark) [reply datalad_registry/models.py:33]: Not particular concern in mind. I just prefer withholding as much internal information as possible from the public as possible.

As an operator, you will, possible always, have access to the command line. Exposing the mapping of dataset URL records and `cache_path` through an API endpoint provides an additional means of access for convenience.

Let me know your final decision on whether you want the command line support alone or both the command line and API endpoint support. I will file an issue correspondingly.

Comment (yarikoptic) [reply datalad_registry/models.py:33]: target use -- command line. I just didn't want to create another app which would interface to the same DB in parallel to the API app.  But if you think it is easier/ok -- go ahead without API component for now.

**Review by yarikoptic — APPROVED** [review]
CLI for getting dataset record for a given cache path I guess could be done later. I have glanced only cursory over the PR -- let's better see it in action!

File: datalad_registry/models.py
Comment (candleindark) [reply datalad_registry/models.py:33]: A new issue is created for this feature. https://github.com/datalad/datalad-registry/issues/165

**Pull request merged**
